feat(trade): cache quotes

shoom3301 opened the pull request:
# Summary

Fixes #5471

Before we used to have quotes caching and it was removed in the recent refactoring.
In this PR I added it back.

# To Test

On Sepolia:
1. Specify WETH -> COW trade
2. Switch token places
- [ ] ER: new quote with sellToken=COW
3. Switch token places again
- [ ] ER: no quote with sellToken=WETH, the price is displayed instantly
4. Select GNO as sell token
- [ ] ER: new quote with sellToken=GNO
5. Select WETH as sell token again
- [ ] ER: no quote with sellToken=WETH, the price is displayed instantly

Also, please check that we don't stuck in "Quote expired" state.
Also, I changed the quote triggering logic. Now we ignore hooks in appData, it might affect Hooks store.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new trade quote management system that enhances the handling of multiple trade quotes.
	- Added functionality for fetching and processing trade quotes more efficiently.
	- Added a new hook for managing trade quote states and behaviors.

- **Refactor**
	- Streamlined token switching and currency selection to ensure a smoother trading experience.
	- Consolidated quote-handling logic to boost performance and system consistency.
	- Simplified state management related to trade quotes for improved clarity.
	- Updated the structure of quote parameters returned from the relevant hooks.

- **Improvements**
	- Enhanced error handling and loading state management for trade quotes.
	- Improved default state handling for unsupported tokens to ensure consistent behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shoom3301 commented on apps/cowswap-frontend/src/modules/advancedOrders/hooks/useAdvancedOrdersActions.ts:
```diff
@@ -21,25 +20,16 @@ export function useAdvancedOrdersActions() {
   const onCurrencySelection = useCallback(
     (field: Field, currency: Currency | null) => {
-      // Reset the output field until we fetch quote for new selected token
-      // This is to avoid displaying wrong amounts in output field
```
This was an execessive amount reset which caused sell amount resetting to 1 when you change sell token

shoom3301 commented on apps/cowswap-frontend/src/modules/advancedOrders/hooks/useAdvancedOrdersActions.ts:
```diff
@@ -61,12 +51,7 @@ export function useAdvancedOrdersActions() {
-
-  const onSwitchTokens = useCallback(() => {
-    onSwitchTokensDefault()
-    resetTradeQuote()
```
Also an excessive quote resetting, this is already handled in `useTradeQuotePolling`

shoom3301 commented on apps/cowswap-frontend/src/modules/volumeFee/state/volumeFeeAtom.ts:
```diff
@@ -56,16 +56,6 @@ const shouldSkipFeeAtom = atom<boolean>((get) => {
   })
-
-  if (isCorrelated) {
-    console.debug('[Volume Fee] Skipping fee for correlated tokens', {
```
Just remove noisy log

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -45,7 +47,7 @@ function removeQuoteMetadata(appData: string | undefined): string | undefined {
-  const { metadata: fullMetadata, ...rest } = decoded
-  const { quote: _, ...metadata } = fullMetadata
+  const { metadata: fullMetadata, ...rest } = decoded as LatestAppDataDocVersion
+  const { quote: _, hooks: __, ...metadata } = fullMetadata
```
This is important.
Hooks update in appData takes some time and it affects quote fetching checks.
In order to not trigger quote fetching to often, we omit `hooks` data.
We need this because if you change sell token from one with permit support to another one with permit support, it will trigger quotes update when we don't need it.
See `AppDataInfoUpdater`

anxolin replied:
But a hook affects the gas costs we receive from the API, right? more gas, more cost fee in sell tokens

shoom3301 replied:
Yes, that's true.
We even discussed it in context of Hooks store: https://cowservices.slack.com/archives/C07FQTUKLJC/p1738850238115259?thread_ts=1738354365.050409&cid=C07FQTUKLJC

Anyway, I will revert this for now, since we already have the same behaviour in prod.

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/services/fetchAndProcessQuote.ts:
```diff
@@ -0,0 +1,42 @@
+  fetchStartTimestamp: number
+}
+
+export async function fetchAndProcessQuote(
```
This was extracted from `useTradeQuotePolling`, no business logic changes here

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/hooks/useTradeQuoteManager.ts:
```diff
@@ -0,0 +1,62 @@
+  onResponse(data: OrderQuoteResponse, requestParams: FeeQuoteParams, fetchStartTimestamp: number): void
+}
+
+export function useTradeQuoteManager(sellTokenAddress: SellTokenAddress | undefined): TradeQuoteManager | null {
```
Extracted the `tradeQuotesAtom` state updates logic in `TradeQuoteManager` in order to make it more clear what are the updates doing.

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/containers/TradeWidget/TradeWidgetUpdaters.tsx:
```diff
@@ -19,28 +19,19 @@ interface TradeWidgetUpdatersProps {
-  const updateQuoteState = useUpdateTradeQuote()
   const priorityTokenAddresses = usePriorityTokenAddresses()
 
-  useEffect(() => {
```
`tradeQuoteStateOverride` was not in use, just deleted it

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/state/tradeQuoteAtom.ts:
```diff
@@ -25,26 +27,36 @@ export const DEFAULT_TRADE_QUOTE_STATE: TradeQuoteState = {
-    }
-  })
-})
+export const tradeQuotesAtom = atom<Record<SellTokenAddress, TradeQuoteState | undefined>>({})
```
Here we have the most important changes. Changes the state from single `TradeQuoteState` to a map.

anxolin replied:
I didn't get why the key is the sell token?

shoom3301 replied:
Basically, I replicated the state from the legacy: https://github.com/cowprotocol/cowswap/blob/release/2024-07-02/apps/cowswap-frontend/src/legacy/state/price/reducer.ts#L26

Sorry, I din't your question, what is supposed to be a key?
If you are worried about other parameters such as amount, kind, etc., the quote refetches when any of them are changed.
The caching only works when you switch token places or change sell token forth and back without amount changes (see "To Test" in the PR description.)

elena-zh reviewed (changes requested):
Hey @shoom3301 , thank you!
Some issues (differences between barn version and the current one): 
1. when a quote is resulted with error, it should be also cached (like it is on Barn). In the current PR a quote is triggered again when flip tokens in the fields
2. In the current PR a quote is triggered when a slippage amount is changed in the settings, while in Bart it is not
3. Refreshing quote state: may hang for too long (up to 1 minute). Might it be possible to force it to run faster?
![hangs](https://github.com/user-attachments/assets/af4d247d-fdf9-46d4-a43f-a4a570c1f1b7)
4. In the current PT caching is applied when flip a token pair on limit orders page while on Barn it is not. I'm not sure about it, maybe we should leave it as it is.. or should fix this? 
5. TWAP: when change a price protection value on the form, a new quote is not triggered while on Bart it is. 
6. Hooks (as you mentioned): adding-removing a hook does not force a quote recalculation on the form, but it should. 

Thanks.

anxolin commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -45,7 +47,7 @@ function removeQuoteMetadata(appData: string | undefined): string | undefined {
 
-  const { metadata: fullMetadata, ...rest } = decoded
-  const { quote: _, ...metadata } = fullMetadata
+  const { metadata: fullMetadata, ...rest } = decoded as LatestAppDataDocVersion
```
nit: maybe you want to do `isAppDataLatestVersion` that just check the property `version` matching `import { LATEST_APP_DATA_VERSION } from '@cowprotocol/app-data'`

shoom3301 replied:
Removed that along with `hooks` omitting

anxolin reviewed (commented):
Thanks for the comments

shoom3301 commented:
@elena-zh thank you!
1. Fixed
2. I can't reproduce that
3. Fixed
4. I think it shoud be consistent with swap
5. Ideally, it should not refetch quote. Price protection technically is a slippage, we should not refetch on slippage changes
6. Removed the logic, now it should trigger quote updating

shoom3301 commented:
@elena-zh 2nd case is fixed now (for eth-flow)

elena-zh commented:
Hey @shoom3301 , everything looks good for me besides 1 issue: quote refreshing still takes a bit longer than I'd expect. F.e. on Barn it takes up to 10 seconds, in the PR it takes 30sec-1 minute :(

shoom3301 commented:
@elena-zh can you check it again please? Just pushed a fix

shoom3301 commented:
@elena-zh I pushed another fix.
Now when you return to the tab it will always refetch the quote.

elena-zh reviewed (approved):
Thank you, works good now!
